src/main/scala/com/cognite/spark/datasource/EventsRelation.scala

```diff
@@ -136,13 +139,111 @@ class EventsRelation(config: RelationConfig)(@transient val sqlContext: SQLConte
+    filter match {
+      case IsNotNull("type") => Seq()
+      case EqualTo("type", value) => Seq(TypeFilter(value.toString))
+      case EqualNullSafe("type", value) => Seq(TypeFilter(value.toString))
+      case In("type", values) => values.map(v => TypeFilter(v.toString))
+      case And(f1, f2) => getTypeFilters(f1) ++ getTypeFilters(f2)
+      case Or(f1, f2) => getTypeFilters(f1) ++ getTypeFilters(f2)
+      // TODO: add support for String filtering using the 'Search for events' endpoint
+      case StringStartsWith("type", value) =>
+        sys.error(
+          s"Filtering using 'string starts with' not allowed for events, attempted for ${value.toString}")
+      case StringEndsWith("type", value) =>
+        sys.error(
+          s"Filtering using 'string ends with' not allowed for events, attempted for ${value.toString}")
+      case StringContains("type", value) =>
+        sys.error(
+          s"Filtering using 'string contains' not allowed for data points, attempted for ${value.toString}")
+      case _ =>
+        Seq()
+    }
+
+  sealed case class SubTypeFilter(subType: String)
+
+  def getSubTypeFilters(filter: Filter): Seq[SubTypeFilter] =
+    filter match {
+      case IsNotNull("subtype") => Seq()
+      case EqualTo("subtype", value) => Seq(SubTypeFilter(value.toString))
+      case EqualNullSafe("subtype", value) => Seq(SubTypeFilter(value.toString))
+      case In("subtype", values) => values.map(v => SubTypeFilter(v.toString))
+      case And(f1, f2) => getSubTypeFilters(f1) ++ getSubTypeFilters(f2)
+      case Or(f1, f2) => getSubTypeFilters(f1) ++ getSubTypeFilters(f2)
+      // TODO: add support for String filtering using the 'Search for events' endpoint
+      case StringStartsWith("subtype", value) =>
+        sys.error(
+          s"Filtering using 'string starts with' not allowed for events, attempted for ${value.toString}")
+      case StringEndsWith("subtype", value) =>
+        sys.error(
+          s"Filtering using 'string ends with' not allowed for events, attempted for ${value.toString}")
+      case StringContains("subtype", value) =>
+        sys.error(
+          s"Filtering using 'string contains' not allowed for data points, attempted for ${value.toString}")
+      case _ =>
+        Seq()
+    }
+  // scalastyle:on cyclomatic.complexity
+  override def buildScan(): RDD[Row] = buildScan(Array.empty, Array.empty)
+
+  override def buildScan(requiredColumns: Array[String], filters: Array[Filter]): RDD[Row] = {
+    val types = filters.flatMap(getTypeFilters).map(_.`type`)
+    val subTypes = filters.flatMap(getSubTypeFilters).map(_.subType)
+
+    val params = Map("type" -> types.toString, "subtype" -> subTypes.toString)
+
+    val rdds: Array[CdpRdd[EventItem]] = for {
+      eventType <- if (types.isEmpty) { Array(None) } else { types }
+      eventSubType <- if (subTypes.isEmpty) { Array(None) } else { subTypes }
+    } yield {
+      val urlWithType =
```

Comment:
`eventType.map(t => listUrl(config).param("type")).getOrElse(listUrl(config))` I think.

Reply:
Or even better, `eventType.fold(listUrl(config))(listUrl(config).param("type", _))` similar to https://github.com/cognitedata/cdp-spark-datasource/blob/0582083db1b0411898357cecc8f53f74a9e375d4/src/main/scala/com/cognite/spark/datasource/CdpRdd.scala#L41

Reply:
Very nice! Thanks